File: strings/number_possible_msg.c
#include <stdio.h>
#include <string.h>


int count_texts(char *s) {
    int memo[5] = {1, 1, 1, 1, 1}, n = strlen(s);
    int max_j;
    for (int i = n - 1; i >= 0; --i) {
        memo[i % 5] = 0;
        if (n < i + (s[i] == '7' || s[i] == '9' ? 4 : 3)) {
            max_j = n;
        }
        else {
            max_j = s[i] == '7' || s[i] == '9' ? 4 : 3;
        }
        for (int j = i; j < max_j && s[i] == s[j]; ++j) {
            memo[i % 5] = (memo[i % 5] + memo[(j + 1) % 5]) % 1000000007;
        }
    }
    return memo[0];
}


int main() {
    printf("%d\n", count_texts("22233"));
    return 0;
}

File: strings/number_possible_msg.py
"""
leetcode 2266: Count number of texts.
"""


def count_texts(s):
    """
    I don't quit understand this one.
    """
    memo = [1] * 5
    for i in reversed(range(len(s))):
        #print("i =", i)
        memo[i % 5] = 0
        #print("memo =", memo)
        if s[i] in '79':
            k = 4
        else:
            k = 3
        max_j = min(len(s), i + k)
        #print("max_j =", max_j)
        for j in range(i, max_j):
            #print("j =", j)
            if s[i] != s[j]:
                #print(f"{s[i]} != {s[j]}, breaking")
                break
            #print(f"memo[{i % 5}] = memo[{i % 5}] + memo[{(j + 1) % 5}]")
            memo[i % 5] = (memo[i % 5] + memo[(j + 1) % 5]) % (1e9 + 7)
            #print("memo =", memo)
            #print()
    return int(memo[0])


def count_texts2(s):
    """
    """
    mod = 1e9 + 7
    memo = [0] * (len(s) + 1)
    memo[0] = 1
    for i in range(1, len(s) + 1):
        memo[i] = (memo[i] + memo[i - 1]) % mod
        if i - 2 >= 0 and s[i - 1] == s[i - 2]:
            memo[i] = (memo[i] + memo[i - 2]) % mod
            if i - 3 >= 0 and s[i - 1] == s[i - 3]:
                memo[i] = (memo[i] + memo[i - 3]) % mod
                if s[i - 1] in "79" and i - 3 >= 0 and s[i - 1] == s[i - 4]:
                    memo[i] = (memo[i] + memo[i - 4]) % mod
    return int(memo[-1])


def test1():
    assert count_texts("22233") == 8
    assert count_texts2("22233") == 8
    print("test 1 successful")


def test2():
    assert count_texts("7777") == 8
    assert count_texts2("7777") == 8
    print("test 2 successful")


def main():
    test1()
    test2()


if __name__ == "__main__":
    main()
